### PR TITLE
[#222] Update `ProtocolMagic` to include `RequiresNetworkMagic`

### LINKED ISSUE #222: Update `ProtocolMagic` to include `RequiresNetworkMagic`

### PR DESCRIPTION
**Pull request by Jimbo4350**
Linked issue: https://iohk.myjetbrains.com/youtrack/issue/CO-354
Closes #222

### CODE REVIEW

File: crypto/src/Cardano/Crypto/ProtocolMagic.hs [hunk 1]
```diff
@@ -1,21 +1,163 @@
...
-  ( ProtocolMagic(..)
+  ( ProtocolMagicId(..)
+  , ProtocolMagic(..)
+  , RequiresNetworkMagic(..)
+  , getProtocolMagic
   )
 where
 
 import Cardano.Prelude
+import Prelude (lookup, String)
```
Comment (ruhatch) [inline crypto/src/Cardano/Crypto/ProtocolMagic.hs]: Please import these from their modules, `Data.List`, `Data.String`, respectively I believe

File: crypto/src/Cardano/Crypto/ProtocolMagic.hs [hunk 2]
```diff
@@ -1,21 +1,163 @@
...
+  --
+  -- See the implementation of `withCoreConfigurations` for more detail on
+  -- how this works.
+  toJSON (ProtocolMagic (ProtocolMagicId ident) _rnm) = toJSON ident
+
+-- Here we default to `RequiresMagic` (what testnets use) if only
+-- a ProtocolMagic identifier is provided.
+instance MonadError SchemaError m => FromJSON m ProtocolMagic where
+  fromJSON = \case
+    (JSNum n) ->
```
Comment (ruhatch) [inline crypto/src/Cardano/Crypto/ProtocolMagic.hs]: I think these brackets are redundant!

File: crypto/src/Cardano/Crypto/ProtocolMagic.hs [hunk 3]
```diff
@@ -1,21 +1,163 @@
...
+-- Canonical JSON instances
+instance Monad m => ToJSON m RequiresNetworkMagic where
+  toJSON RequiresNoMagic = pure (JSString "RequiresNoMagic")
+  toJSON RequiresMagic   = pure (JSString "RequiresMagic")
 
-deriving instance ToJSON ProtocolMagic
-deriving instance FromJSON ProtocolMagic
+instance MonadError SchemaError m => FromJSON m RequiresNetworkMagic where
+  fromJSON = \case
+    (JSString "RequiresNoMagic") -> pure RequiresNoMagic
```
Comment (ruhatch) [inline crypto/src/Cardano/Crypto/ProtocolMagic.hs]: And these too

File: crypto/test/Test/Cardano/Crypto/Arbitrary.hs [hunk 4]
```diff
@@ -53,10 +54,18 @@ import Cardano.Crypto.Signing.Redeem
   (RedeemPublicKey, RedeemSecretKey, RedeemSignature, redeemKeyGen, redeemSign)
 
 import Test.Cardano.Crypto.Arbitrary.Unsafe ()
-import Test.Cardano.Crypto.Dummy (dummyProtocolMagic)
+import Test.Cardano.Crypto.Dummy (dummyProtocolMagicId)
 
 
-deriving instance Arbitrary ProtocolMagic
+instance Arbitrary ProtocolMagic where
```
Comment (ruhatch) [inline crypto/test/Test/Cardano/Crypto/Arbitrary.hs:60]: Not sure these will be necessary - try deleting them and recompiling to see if there's any QuickCheck stuff using them

Comment (ruhatch) [reply crypto/test/Test/Cardano/Crypto/Arbitrary.hs:60]: Oh actually, I don't think I finished removing QuickCheck here

File: crypto/test/Test/Cardano/Crypto/Dummy.hs [hunk 5]
```diff
@@ -2,10 +2,15 @@
...
   )
 where
 
-import Cardano.Crypto (ProtocolMagic(..))
+import Cardano.Crypto
+  (ProtocolMagic(..), ProtocolMagicId(..), RequiresNetworkMagic(..))
 
 dummyProtocolMagic :: ProtocolMagic
-dummyProtocolMagic = ProtocolMagic 55550001
+dummyProtocolMagic = ProtocolMagic (ProtocolMagicId 55550001) RequiresMagic
```
Comment (ruhatch) [inline crypto/test/Test/Cardano/Crypto/Dummy.hs]: Might as well do:
```suggestion
dummyProtocolMagic = ProtocolMagic dummyProtocolMagicId RequiresMagic
```

File: src/Cardano/Chain/Block/Header.hs [hunk 6]
```diff
@@ -161,7 +161,7 @@ instance B.Buildable Header where
 instance Bi Header where
   encode h =
     encodeListLen 5
-      <> encode (getProtocolMagic (headerProtocolMagic h))
+      <> encode (unProtocolMagicId $ (headerProtocolMagicId h))
```
Comment (ruhatch) [inline src/Cardano/Chain/Block/Header.hs]: Could be better to `deriving Bi` on `ProtocolMagicId` type using `GeneralizedNewtypeDeriving` so you get the unwrapping for free

File: src/Cardano/Chain/Block/Header.hs [hunk 7]
```diff
@@ -175,7 +175,7 @@ decodeAHeader = do
     annotatedDecoder $ do
       enforceSize "Header" 5
       (,,,,)
-        <$> (ProtocolMagic <$> decode)
+        <$> (ProtocolMagicId <$> decode)
```
Comment (ruhatch) [inline src/Cardano/Chain/Block/Header.hs]: And then you could change this too

File: src/Cardano/Chain/Delegation/Validation.hs [hunk 8]
```diff
@@ -127,7 +132,7 @@ scheduleCertificate config slot d ss cert = do
     `orThrowError` SchedulingMultipleDelegationsForSlot slot delegator
 
   -- Check that the delegation certificate is valid
-  validateProxySecretKey (configProtocolMagic config) cert
+  validateProxySecretKey (getProtocolMagicId $ configProtocolMagic config) cert
```
Comment (ruhatch) [inline src/Cardano/Chain/Delegation/Validation.hs]: I would define a `configProtocolMagicId :: Config -> ProtocolMagicId` as this is used in a couple of places

Comment (ruhatch) [reply src/Cardano/Chain/Delegation/Validation.hs]: Define that alongside `Config` of course, not here

File: src/Cardano/Chain/Genesis/Data.hs [hunk 9]
```diff
@@ -68,7 +69,7 @@ instance Monad m => ToJSON m GenesisData where
     , ( "protocolConsts"
       , mkObject
         [ ("k"            , pure . JSNum . fromIntegral $ gdK gd)
-        , ("protocolMagic", toJSON . getProtocolMagic $ gdProtocolMagic gd)
+        , ("protocolMagic", toJSON . unProtocolMagicId.  getProtocolMagicId $ gdProtocolMagic gd)
```
Comment (ruhatch) [inline src/Cardano/Chain/Genesis/Data.hs]: Didn't we define `getProtocolMagic` in terms of these two?

**Review by ruhatch — CHANGES_REQUESTED** [review]
Looks good! Glad it wasn't as extensive as the `cardano-sl` implementation!

Just a few small changes to be made :)

File: crypto/src/Cardano/Crypto/ProtocolMagic.hs (hunk 1 above)
Comment (Jimbo4350) [reply crypto/src/Cardano/Crypto/ProtocolMagic.hs]: Forgot I did that!

File: crypto/src/Cardano/Crypto/ProtocolMagic.hs [hunk 10]
```diff
@@ -1,21 +1,171 @@
...
 
-import Data.Aeson (FromJSON(..), ToJSON(..))
+import qualified Data.Aeson as A
+import Data.Aeson ((.:), (.=))
+import Data.Aeson.Types (typeMismatch)
+import Data.List (lookup)
+import Data.String (String)
+import Text.JSON.Canonical (FromJSON(..), JSValue(..), ToJSON(..), expected)
 
+import Cardano.Binary.Class (Bi)
```
Comment (ruhatch) [inline crypto/src/Cardano/Crypto/ProtocolMagic.hs:31]: Can you put two blank lines after imports here?

File: crypto/src/Cardano/Crypto/Signing/Tag.hs [hunk 11]
```diff
@@ -69,4 +69,4 @@ signTag protocolMagic = \case
   -- "\x08" was used for SignMainBlockLight, but was never used in mainnet
   SignMainBlockHeavy -> "\x09" <> network
   SignProxySK        -> "\x0a" <> network
-  where network = Bi.serialize' (getProtocolMagic protocolMagic)
+  where network = Bi.serialize' (unProtocolMagicId $ protocolMagic)
```
Comment (ruhatch) [inline crypto/src/Cardano/Crypto/Signing/Tag.hs:72]: You should be able to use the `ProtocolMagicId` instance here I think
```suggestion
  where network = Bi.serialize' protocolMagic
```

**Review by ruhatch — APPROVED** [review]
A couple of small changes, then feel free to merge!

**Pull request merged**